user/tests/wiring/no_fixture/thread.cpp

```diff
@@ -203,7 +203,11 @@ test(THREAD_07_particle_process_behavior_when_threading_enabled)
     assertTrue(Particle.connected());
 
     assertEqual((int)test_val_fn1, 0);
-    Particle.process();
+    for (int i = 0; i < 20; i++) {
+        // The increment task that was added to the Application Queue via invoke_async is not guarenteed to be the first and only item in the Application Thread Queue
+        // Call process() repeatedly in order to ensure that all potential messages in the queue are consumed, including the increment function
+        Particle.process();    
```

Comment:
We probably need to look at the other tests that use `Particle.process()` as well and add a similar thing in there.

How about we make this a bit more deterministic? We can extend `spark_process()` to return a bool/int to indicate whether there are any more events to process (void -> some type is fine as far as ABI goes) in the queue at least in the case of threaded mode and at least in the tests we can make a decision based on that.

Reply:
I agree there are likely other tests that rely on this behavior that might fail in this scenario. I dont know if we want to find and change all of them right now.

If we want to improve this, we will have to change `Particle.process()` itself right? 
This means changing the `CloudClass::process` method to something like
```
    static bool process(void) {
            application_checkin();
            return spark_process();
    }
```
Or if we want it to process everything in the queue:
```
    static void process(void) {
            application_checkin();
            bool process = false;
            do {
                process = spark_process();
            } while (process);
    }
```
Or do this inside `spark_process()` anyways without having to change dynalib:
```
void spark_process(void)
{
    // application thread will pump application messages
#if PLATFORM_THREADING
    if (APPLICATION_THREAD_CURRENT()) {
        if (system_thread_get_state(NULL)) {
            bool process = false;
            do {
                process = ApplicationThread.process();
            } while (process);
        } else {
            Spark_Idle_Events(true);
        }
    }
#else
    // run the background processing loop, and specifically also pump cloud events
    Spark_Idle_Events(true);
#endif // PLATFORM_THREADING
}
```

I think any fix to this problem shouldnt involve having to loop over `Particle.process()` everywhere it is used. But this will change the system task behavior as well....

Reply:
What I'm proposing 'for now' is:
```diff
diff --git a/system/inc/system_dynalib_cloud.h b/system/inc/system_dynalib_cloud.h
index 107d824e8..cb3464975 100644
--- a/system/inc/system_dynalib_cloud.h
+++ b/system/inc/system_dynalib_cloud.h
@@ -37,7 +37,7 @@ DYNALIB_BEGIN(system_cloud)
 
 DYNALIB_FN(0, system_cloud, spark_variable, bool(const char*, const void*, Spark_Data_TypeDef, spark_variable_t*))
 DYNALIB_FN(1, system_cloud, spark_function, bool(const char*, p_user_function_int_str_t, void*))
-DYNALIB_FN(2, system_cloud, spark_process, void(void))
+DYNALIB_FN(2, system_cloud, spark_process, bool(void))
 DYNALIB_FN(3, system_cloud, spark_cloud_flag_connect, void(void))
 DYNALIB_FN(4, system_cloud, spark_cloud_flag_disconnect, void(void))
 DYNALIB_FN(5, system_cloud, spark_cloud_flag_connected, bool(void))
diff --git a/system/src/system_cloud.cpp b/system/src/system_cloud.cpp
index c7153339d..0e348dc8a 100644
--- a/system/src/system_cloud.cpp
+++ b/system/src/system_cloud.cpp
@@ -257,13 +257,13 @@ int spark_cloud_disconnect(const spark_cloud_disconnect_options* options, void*
     return 0;
 }
 
-void spark_process(void)
+bool spark_process(void)
 {
        // application thread will pump application messages
 #if PLATFORM_THREADING
     if (APPLICATION_THREAD_CURRENT()) {
         if (system_thread_get_state(NULL)) {
-            ApplicationThread.process();
+            return ApplicationThread.process();
         } else {
             Spark_Idle_Events(true);
         }
@@ -272,6 +272,7 @@ void spark_process(void)
     // run the background processing loop, and specifically also pump cloud events
     Spark_Idle_Events(true);
 #endif // PLATFORM_THREADING
+    return false;
 }
 
 String spark_deviceID(void)
diff --git a/wiring/inc/spark_wiring_cloud.h b/wiring/inc/spark_wiring_cloud.h
index f4587e8c2..12ef74437 100644
--- a/wiring/inc/spark_wiring_cloud.h
+++ b/wiring/inc/spark_wiring_cloud.h
@@ -369,9 +369,9 @@ public:
         spark_cloud_flag_connect();
     }
     static void disconnect(const CloudDisconnectOptions& options = CloudDisconnectOptions());
-    static void process(void) {
-               application_checkin();
-               spark_process();
+    static bool process(void) {
+        application_checkin();
+        return spark_process();
     }
     static String deviceID(void) { return SystemClass::deviceID(); }
 
```

That way we can for now 'fix' the tests to loop around Particle.process() until it stops saying `true` which means that the queue has been finally flushed, and in the future we can also potentially modify the behavior of `Particle.process()` specifically, or whatever else again based on the queue state.

Reply:
ok, thats what I was thinking originally. ill add that and we can loop on `Particle.process()` in this test, which is better than just polling an arbitrary amount.